File: public/cpp/090-k-closest-points-to-origin.cpp
/*
 * Problem: K Closest Points to Origin (LeetCode 973)
 * Link: https://leetcode.com/problems/k-closest-points-to-origin/
 * Difficulty: Medium
 * Category: Heaps and Priority Queues
 *
 * Description:
 * Given an array of points where points[i] = [xi, yi] represents a point on
 * the X-Y plane and an integer k, return the k closest points to the origin
 * (0, 0). The distance from (0,0) to (x,y) is sqrt(x²+y²). You may return
 * the answer in any order. The answer is guaranteed to be unique.
 *
 * Example 1: points=[[1,3],[-2,2]] k=1  → [[-2,2]]
 *   dist([1,3])=√10 > dist([-2,2])=√8 → [-2,2] is closer
 * Example 2: points=[[3,3],[5,-1],[-2,4]] k=2 → [[3,3],[-2,4]]
 *
 * Constraints:
 * - 1 <= k <= points.length <= 10^4
 * - -10^4 <= xi, yi <= 10^4
 */

#include <vector>
#include <algorithm>
#include <queue>
#include <cmath>
#include <iostream>
using namespace std;

using Point = vector<int>;
auto dist2 = [](const Point& p){ return p[0]*p[0] + p[1]*p[1]; };

/*
 * APPROACH 1: SORT BY DISTANCE²
 *
 * Sort all points by squared distance, return first k.
 * Avoids sqrt; easy to implement.
 *
 * Time:  O(n log n)
 * Space: O(1) in-place
 */
class Solution1 {
public:
    vector<Point> kClosest(vector<Point>& points, int k) {
        sort(points.begin(), points.end(),
             [](const Point& a, const Point& b){ return dist2(a) < dist2(b); });
        return { points.begin(), points.begin() + k };
    }
};

/*
 * APPROACH 2: MAX HEAP OF SIZE K
 *
 * Maintain a max-heap of size k keyed by distance².
 * For each point: if heap size < k push; else if dist < heap top, replace.
 * Heap top is always the farthest of the k closest seen so far.
 *
 * Time:  O(n log k)
 * Space: O(k)
 */
class Solution2 {
public:
    vector<Point> kClosest(vector<Point>& points, int k) {
        // Max-heap: (dist², index)
        priority_queue<pair<int,int>> pq;
        for (int i = 0; i < (int)points.size(); i++) {
            pq.push({dist2(points[i]), i});
            if ((int)pq.size() > k) pq.pop();
        }
        vector<Point> res;
        while (!pq.empty()) {
            res.push_back(points[pq.top().second]);
            pq.pop();
        }
        return res;
    }
};

/*
 * APPROACH 3: QUICKSELECT (AVERAGE O(n))
 *
 * Partition around pivot until the first k elements are the k closest.
 * nth_element in STL does this in O(n) average.
 *
 * Time:  O(n) average, O(n²) worst
 * Space: O(1)
 */
class Solution3 {
public:
    vector<Point> kClosest(vector<Point>& points, int k) {
        nth_element(points.begin(), points.begin() + k, points.end(),
                    [](const Point& a, const Point& b){ return dist2(a) < dist2(b); });
        return { points.begin(), points.begin() + k };
    }
};

/*
 * APPROACH 4: PARTIAL SORT (O(n log k))
 *
 * partial_sort guarantees first k are sorted by distance.
 * Cleaner than full sort; slightly better for k << n.
 *
 * Time:  O(n log k)
 * Space: O(1)
 */
class Solution4 {
public:
    vector<Point> kClosest(vector<Point>& points, int k) {
        partial_sort(points.begin(), points.begin() + k, points.end(),
                     [](const Point& a, const Point& b){ return dist2(a) < dist2(b); });
        return { points.begin(), points.begin() + k };
    }
};

/*
 * APPROACH 5: MANUAL QUICKSELECT
 *
 * In-place partition-based selection: find the pivot position such that
 * all points left of pivot are closer than all points right of it.
 *
 * Time:  O(n) average
 * Space: O(log n) recursion stack
 */
class Solution5 {
    int partition(vector<Point>& pts, int lo, int hi) {
        int pivot = dist2(pts[hi]), i = lo;
        for (int j = lo; j < hi; j++)
            if (dist2(pts[j]) <= pivot) swap(pts[i++], pts[j]);
        swap(pts[i], pts[hi]);
        return i;
    }
    void quickselect(vector<Point>& pts, int lo, int hi, int k) {
        if (lo >= hi) return;
        int p = partition(pts, lo, hi);
        if (p == k) return;
        if (p < k) quickselect(pts, p+1, hi, k);
        else quickselect(pts, lo, p-1, k);
    }
public:
    vector<Point> kClosest(vector<Point>& points, int k) {
        quickselect(points, 0, (int)points.size()-1, k);
        return { points.begin(), points.begin() + k };
    }
};

void print(const vector<Point>& v) {
    cout << "[";
    for (int i=0;i<(int)v.size();i++) {
        cout << "[" << v[i][0] << "," << v[i][1] << "]";
        if (i+1<(int)v.size()) cout << ",";
    }
    cout << "]";
}

int main() {
    auto run = [](vector<Point> pts, int k) {
        cout << "k=" << k << "  ";
        { Solution1 s; auto r=s.kClosest(pts,k); cout<<"Sort:    "; print(r); cout<<"\n"; }
        { vector<Point> p2=pts; Solution2 s; auto r=s.kClosest(p2,k); cout<<"MaxHeap: "; print(r); cout<<"\n"; }
        { vector<Point> p3=pts; Solution3 s; auto r=s.kClosest(p3,k); cout<<"QSel:    "; print(r); cout<<"\n"; }
        { vector<Point> p4=pts; Solution4 s; auto r=s.kClosest(p4,k); cout<<"PSort:   "; print(r); cout<<"\n"; }
        { vector<Point> p5=pts; Solution5 s; auto r=s.kClosest(p5,k); cout<<"ManQSel: "; print(r); cout<<"\n"; }
        cout << "\n";
    };
    run({{1,3},{-2,2}}, 1);
    run({{3,3},{5,-1},{-2,4}}, 2);
    run({{0,1},{1,0}}, 2);
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (Sort):           Time O(n log n)   Space O(1)
 * Approach 2 (Max Heap k):     Time O(n log k)   Space O(k)
 * Approach 3 (nth_element):    Time O(n) avg     Space O(1)
 * Approach 4 (partial_sort):   Time O(n log k)   Space O(1)
 * Approach 5 (Manual QSelect): Time O(n) avg     Space O(log n)
 *
 * KEY INSIGHT: Use squared distance to avoid sqrt.
 */

// Made with Bob

File: src/data/tutorials/090-k-closest-points-to-origin.js
/**
 * Tutorial content for #090 — K Closest Points to Origin
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an array of points on a 2D plane, return the k closest points to the origin (0, 0). Distance is measured as Euclidean distance: √(x² + y²). You may return the answer in any order. There is guaranteed to be a unique answer (no ties at the kth position).`,
    example: `points = [[1,3],[-2,2]], k = 1\nDist² of [1,3]:  1+9  = 10\nDist² of [-2,2]: 4+4  = 8  ← smaller\n✅ Answer: [[-2,2]]`,
    keyInsight: `You don't need to compute actual Euclidean distances (no square root needed) because comparisons only require relative distances, and x²+y² preserves that order. Use a max-heap of size k: discard the farthest point whenever the heap exceeds k.`,
  },

  approaches: {
    'Sort by Distance²': {
      intuition: `Compute x²+y² for every point, sort all points by that value, and return the first k. Simple, correct, and fast enough for most inputs.`,
      steps: [
        `Sort points by (x*x + y*y) ascending.`,
        `Return the first k points.`,
      ],
      example: `points=[[1,3],[-2,2],[5,0]], k=2\nDist²: 10, 8, 25\nSorted: [[-2,2],[1,3],[5,0]]\nFirst 2: [[-2,2],[1,3]] ✅`,
      keyInsight: `O(n log n) time, O(1) extra space (in-place sort). The easiest correct solution. Great starting point in an interview before optimizing.`,
    },

    'Max Heap of Size K': {
      intuition: `Maintain a max-heap of size k keyed by squared distance. For each point, add it to the heap. If the heap grows beyond k, remove the farthest point (the max). After all points are processed, the heap holds the k closest points.`,
      steps: [
        `Create a max-heap (by x²+y²).`,
        `For each point:`,
        `  Add it to the heap.`,
        `  If heap.size > k: remove the farthest (max).`,
        `Return all points remaining in the heap.`,
      ],
      example: `points=[[1,3],[-2,2],[5,0],[0,1]], k=2\n\nAdd [1,3](d²=10): heap:[(10,[1,3])]\nAdd [-2,2](d²=8): heap:[(10,[1,3]),(8,[-2,2])]\nAdd [5,0](d²=25): heap:[(25,[5,0]),(10,[1,3]),(8,[-2,2])] size=3>2 → remove max(25).\n  heap:[(10,[1,3]),(8,[-2,2])]\nAdd [0,1](d²=1): heap:[(10,[1,3]),(8,[-2,2]),(1,[0,1])] size=3>2 → remove max(10).\n  heap:[(8,[-2,2]),(1,[0,1])]\nResult: [[-2,2],[0,1]] ✅`,
      keyInsight: `O(n log k) time, O(k) space. Better than full sort when k << n. This is the standard "top k" heap pattern.`,
    },

    'Quickselect': {
      intuition: `Quickselect finds the kth smallest element in O(n) average time without full sorting. Pick a pivot distance, partition points into "closer" and "farther" groups, and recurse on the side containing the kth position. Return all points on the "closer" side once the pivot lands at position k.`,
      steps: [
        `If k == n: return all points.`,
        `Randomly pick a pivot point.`,
        `Partition: points closer than pivot go to the left, farther go right.`,
        `If left.size == k: return left.`,
        `If left.size > k: recurse Quickselect(left, k).`,
        `Else: recurse Quickselect(right, k - left.size).`,
      ],
      example: `points=[[1,3],[-2,2],[5,0],[0,1]], k=2\nPick pivot [-2,2] (d²=8).\nLeft (d²≤8): [[-2,2],[0,1]]  Right: [[1,3],[5,0]]\nleft.size=2 == k=2 → return [[-2,2],[0,1]] ✅`,
      keyInsight: `O(n) average, O(n²) worst-case time. O(1) extra space. Fastest on average for large arrays — used in practice when you need raw speed without the O(n log n) sort overhead.`,
    },

    'Partial Sort': {
      intuition: `Use a language built-in that can do a partial sort — Java's Arrays.sort with a custom comparator, Python's heapq.nsmallest, C++'s partial_sort or nth_element. The idea is identical to Sort by Distance² but using a library function that may be optimized for this case.`,
      steps: [
        `Use heapq.nsmallest(k, points, key=lambda p: p[0]**2 + p[1]**2) in Python.`,
        `Or Arrays.sort(points, Comparator.comparingInt(p -> p[0]*p[0] + p[1]*p[1])) and take first k.`,
        `Or partial_sort / nth_element in C++.`,
      ],
      example: `points=[[1,3],[-2,2],[5,0]], k=1\nheapq.nsmallest(1, points, key=dist²)\n→ dist²: 10, 8, 25\n→ smallest 1: [[-2,2]] ✅`,
      keyInsight: `O(n log k) for heapq.nsmallest, O(n log n) for full sort. Leverages highly optimized library internals. Pragmatic choice in a time-pressured coding environment.`,
    },

    'Manual Quickselect': {
      intuition: `Same as Quickselect above, but implemented entirely by hand using in-place two-pointer partitioning on the original array — similar to how Quicksort's partition works. Avoids creating sub-arrays, modifying the input in-place.`,
      steps: [
        `Select a random pivot index; swap it to the end.`,
        `left pointer = start, right pointer = end-1.`,
        `Move all points with d² ≤ pivot's d² to the left using two-pointer swap.`,
        `Place pivot at left pointer position.`,
        `Compare pivot's position to k: recurse on the appropriate half.`,
      ],
      example: `points=[[1,3],[-2,2],[5,0],[0,1]], k=2 (0-indexed: want positions 0 and 1)\nPivot=[1,3](d²=10), swap to end: [[−2,2],[5,0],[0,1],[1,3]]\nPartition (d²≤10): [-2,2](8≤10✓), [5,0](25>10✗ stay right), [0,1](1≤10✓)\nAfter partition+place pivot: [[-2,2],[0,1],[1,3],[5,0]]\nPivot at index 2. k=2 → left 2 elements are the answer.\nReturn [[-2,2],[0,1]] ✅`,
      keyInsight: `O(n) average, O(1) extra space (in-place). The most memory-efficient solution. Tricky to implement correctly under pressure — the library Quickselect or heap approach is safer in interviews.`,
    },
  },
}
